packages/vite/src/node/server/index.ts

```diff
@@ -371,6 +371,12 @@ export async function createServer(
       }
     },
     listen(port?: number, isRestart?: boolean) {
+      if (port !== undefined) {
+        server.config.inlineConfig.server = {
+          ...(config.inlineConfig.server || {}),
+          port,
+        };
+      }
```

Comment:
I don't think that we should modify the inline config here. IIUC, this will still expose the same issue when the port is occupied and you aren't using strictPort.

I think the solution is to get the current port in the `restartServer` function before closing the server and passing that to listen here https://github.com/vitejs/vite/blob/1e711c0ef639511240523cb99c2bcc901b22b3e6/packages/vite/src/node/server/hmr.ts#L489

For reference, this is another PR that is also affecting the same function because of another issue https://github.com/vitejs/vite/pull/4374